### DISCUSSION TITLE
BIRP Lighting

### DISCUSSION THREAD

**Discussion by Revmatek** [crest-ocean/crest#1074] [General]
https://crest.readthedocs.io/en/latest/user/configuration.html?rp=birp

Crest BIRP does not support additional lights due to bugs in the pipeline and performance concerns. Please see [#382](https://github.com/wave-harmonic/crest/pull/382) and [#383](https://github.com/wave-harmonic/crest/pull/383) for more details.

Early on when I was evaluating which to use this, LUX, various other time savers, Lighting was one of the deal breakers (for right or wrong). At some point I evaluated Aura 2 and modified the Ocean.shader to have custom pass that adds volumetric lighting which at the time was a compromise for no light component support at all, other than directional lighting. This came at the cost of performance obviously but it was a desired trade off, something the user can turn off in settings for lower end machines, now I am no longer using Aura 2 due to unrelated reasons / requirements, but it proved to me it was possible to have a good lighting feature for this. At the time I also saw the 2 experimental branches that added lighting and I understood the concerns. But why not have this as a toggleable feature? Experimental? Is there still a show stopping bug? There must be some sort of workaround because it seems to work fluidly for similar projects some examples that do not meet my needs for other reasons but prove to me it is an obtainable goal:
https://youtu.be/uiEgukNGxrA (point-light)
https://youtu.be/1zx2hFxH7BI (spot-lights)

There are some other examples for URP, but I am not concerned about that.

Planar reflections works great and has a noticeable performance hit (which is expected), I think it would be totally acceptable to have additional lights in a similar manner. I tried briefly to adapt the branches to the current baselines but did not have the time to proceed to a satisfactory level. The screenshots I saw would have been more than satisfactory for my use case, especially if it could be turned off when not needed. Some of the other time saving ocean / water projects out there (lacking many features) support lighting in this manner (or at least assert to). Is this something that can be put on a roadmap / future plan?  I really feel like if Crest supported additional lighting there would be little reason to use anything else in most cases baring uses that required no more than a puddle of water or a pond you can't actually swim in but is more for scenic purposes, namely a functional ocean / body of water.

**Comment by magenta404** [comment]
There is still work to be done for both of those branches as they are very outdated with the primary branch and incomplete. Unfortunately, available resources has been lacking and finishing and maintaining feature is not a priority for the GH version.

Crest will support additional lights through Shader Graph. Work is underway to release a package on the Unity Asset Store which supports all three render pipelines.

**Reply by Revmatek** [reply]
That is understandable. I assume that may mean URP/HDRP only for a future feature. So would those shader graphs be portable to amplify shader?  https://assetstore.unity.com/packages/tools/visual-scripting/amplify-shader-editor-68570

**Reply by magenta404** [reply]
URP/HDRP already have it. SG now supports BIRP too so it should also receive the feature once the asset is released. I haven't used ASE so I cannot comment but I don't believe they are compatible with each other.

**Reply by Revmatek** [reply]
Ok cool. I'll keep an eye out for the asset. 

Good to know about the SG support. My understanding of ASE, with my limited usage of it is that it is just another shader graph UI that makes shaders for BIRP in the end. You can take shaders made in Unreal Engine and port them into a similar graph so I image if a graph exists, it can be ported baring lacking support for specific features, but most of the stuff for URP is there I believe, I cannot say what sort of pitfalls are in there, but if SG works for BIRP now, it may be a none issue. Thanks for the info.
